cmd/testrunner.go

```diff
@@ -58,345 +56,842 @@ These tests allow you to test different configuration options and the policies t
-					return cobraext.FlagParsingError(err, cobraext.DataStreamsFlagName)
-				}
+		if failOnMissing && len(testFolders) == 0 {
+			if len(dataStreams) > 0 {
+				return fmt.Errorf("no %s tests found for %s data stream(s)", testType, strings.Join(dataStreams, ","))
 			}
+			return fmt.Errorf("no %s tests found", testType)
+		}
+	} else {
+		testFolders, err = testrunner.FindTestFolders(packageRootPath, nil, testType)
+		if err != nil {
+			return fmt.Errorf("unable to determine test folder paths: %w", err)
+		}
+		if failOnMissing && len(testFolders) == 0 {
+			return fmt.Errorf("no %s tests found", testType)
+		}
+	}
 
-			if runner.TestFolderRequired() {
-				testFolders, err = testrunner.FindTestFolders(packageRootPath, dataStreams, testType)
-				if err != nil {
-					return fmt.Errorf("unable to determine test folder paths: %w", err)
-				}
-			} else {
-				testFolders, err = testrunner.AssumeTestFolders(packageRootPath, dataStreams, testType)
-				if err != nil {
-					return fmt.Errorf("unable to assume test folder paths: %w", err)
-				}
-			}
+	ctx, stop := signal.Enable(cmd.Context(), logger.Info)
+	defer stop()
 
-			if failOnMissing && len(testFolders) == 0 {
-				if len(dataStreams) > 0 {
-					return fmt.Errorf("no %s tests found for %s data stream(s)", testType, strings.Join(dataStreams, ","))
-				}
-				return fmt.Errorf("no %s tests found", testType)
-			}
-		} else {
-			_, pkg := filepath.Split(packageRootPath)
+	kibanaClient, err := stack.NewKibanaClientFromProfile(profile)
+	if err != nil {
+		return fmt.Errorf("can't create Kibana client: %w", err)
+	}
 
-			if runner.TestFolderRequired() {
-				testFolders, err = testrunner.FindTestFolders(packageRootPath, nil, testType)
-				if err != nil {
-					return fmt.Errorf("unable to determine test folder paths: %w", err)
-				}
-				if failOnMissing && len(testFolders) == 0 {
-					return fmt.Errorf("no %s tests found", testType)
-				}
-			} else {
-				testFolders = []testrunner.TestFolder{
-					{
-						Package: pkg,
-					},
-				}
-			}
+	esClient, err := stack.NewElasticsearchClientFromProfile(profile)
+	if err != nil {
+		return fmt.Errorf("can't create Elasticsearch client: %w", err)
+	}
+	err = esClient.CheckHealth(ctx)
+	if err != nil {
+		return err
+	}
+
+	if runTearDown || runTestsOnly {
+		if variantFlag != "" {
+			return fmt.Errorf("variant flag cannot be set with --tear-down or --no-provision")
+		}
+	}
+
+	if runSetup || runTearDown || runTestsOnly {
+		// variant flag is not checked here since there are packages that do not have variants
+		if len(testFolders) != 1 {
+			return fmt.Errorf("wrong number of test folders (expected 1): %d", len(testFolders))
 		}
 
-		deferCleanup, err := cmd.Flags().GetDuration(cobraext.DeferCleanupFlagName)
+		cmd.Printf("Running tests per stages (technical preview)\n")
+	}
+
+	var results []testrunner.TestResult
+	for _, folder := range testFolders {
+		r, err := testrunner.Run(ctx, testType, testrunner.TestOptions{
```

Comment:
Now that we have specific commands for each test runner, maybe we can call here the specific runner, with an specific set of options? This would also allow to remove the registry of tests.

Actually now we wouldn't even need the test runner interface, right? :thinking:

Reply:
It looks like this could be possible, I'll check this out in a follow-up PR.